File: Examples/example/atm/FraudMonitorTest.cs
using System;

using NUnit.Framework;

using net.esper.client;
using net.esper.client.time;
using net.esper.compat;
using net.esper.support.util;
using net.esper.events;

namespace net.esper.example.atm
{
	[TestFixture]
	public class FraudMonitorTest
	{
		private EPServiceProvider epService;
		private SupportUpdateListener listener;

		[SetUp]
	    public void setUp()
    	{
	        Configuration configuration = new Configuration();
	        configuration.AddEventTypeAlias("FraudWarning", typeof(FraudWarning).FullName);
	        configuration.AddEventTypeAlias("Withdrawal", typeof(Withdrawal).FullName);
	
	        epService = EPServiceProviderManager.GetProvider("FraudMonitorTest", configuration);
	        epService.EPRuntime.SendEvent(new TimerControlEvent(TimerControlEvent.ClockTypeEnum.CLOCK_EXTERNAL));
	
	        this.listener = new SupportUpdateListener();
	        new FraudMonitor(epService, listener.Update);
	    }
	    
		[Test]
		public void testJoin()
	    {
	        const string FRAUD_TEXT = "card reported stolen";
	
	        sendWithdrawal(1001, 100);
	        sendFraudWarn(1004, FRAUD_TEXT);
	        sendWithdrawal(1001, 60);
	        sendWithdrawal(1002, 400);
	        sendWithdrawal(1001, 300);
	
	        Assert.IsFalse(listener.getAndClearIsInvoked());
	
	        sendWithdrawal(1004, 100);
	        Assert.IsTrue(listener.getAndClearIsInvoked());
	
	        Assert.AreEqual(1, listener.LastNewData.Length);
	        EventBean _event = listener.LastNewData[0];
	        Assert.AreEqual(1004L, _event["AccountNumber"]);
	        Assert.AreEqual(FRAUD_TEXT, _event["Warning"]);
	        Assert.AreEqual(100, _event["Amount"]);
	        Assert.IsTrue( ((long) _event["Timestamp"]) > (DateTimeHelper.CurrentTimeMillis - 100));
	        Assert.AreEqual("withdrawlFraudWarn", _event["Descr"]);
	    }
	
	    private void sendWithdrawal(long acctNum, int amount)
	    {
	        Withdrawal eventBean = new Withdrawal(acctNum, amount);
	        epService.EPRuntime.SendEvent(eventBean);
	    }
	
	    private void sendFraudWarn(long acctNum, String text)
	    {
	        FraudWarning eventBean = new FraudWarning(acctNum, text);
	        epService.EPRuntime.SendEvent(eventBean);
	    }
	}
}

File: trunk/examples/src/test/java/net/esper/example/atm/FraudMonitorTest.java
package net.esper.example.atm;

import junit.framework.TestCase;
import net.esper.client.EPServiceProvider;
import net.esper.client.EPServiceProviderManager;
import net.esper.client.Configuration;
import net.esper.client.time.TimerControlEvent;
import net.esper.example.support.SupportUpdateListener;
import net.esper.event.EventBean;

public class FraudMonitorTest extends TestCase
{
    private EPServiceProvider epService;
    private SupportUpdateListener listener;

    public void setUp()
    {
        Configuration configuration = new Configuration();
        configuration.addEventTypeAlias("FraudWarning", FraudWarning.class.getName());
        configuration.addEventTypeAlias("Withdrawal", Withdrawal.class.getName());

        epService = EPServiceProviderManager.getProvider("FraudMonitorTest", configuration);
        epService.getEPRuntime().sendEvent(new TimerControlEvent(TimerControlEvent.ClockType.CLOCK_EXTERNAL));

        this.listener = new SupportUpdateListener();
        new FraudMonitor(epService, listener);
    }
    
    public void testJoin()
    {
        final String FRAUD_TEXT = "card reported stolen";

        sendWithdrawal(1001, 100);
        sendFraudWarn(1004, FRAUD_TEXT);
        sendWithdrawal(1001, 60);
        sendWithdrawal(1002, 400);
        sendWithdrawal(1001, 300);

        assertFalse(listener.getAndClearIsInvoked());

        sendWithdrawal(1004, 100);
        assertTrue(listener.getAndClearIsInvoked());

        assertEquals(1, listener.getLastNewData().length);
        EventBean event = listener.getLastNewData()[0];
        assertEquals(1004L, event.get("accountNumber"));
        assertEquals(FRAUD_TEXT, event.get("warning"));
        assertEquals(100, event.get("amount"));
        assertTrue( ((Long) event.get("timestamp")) > (System.currentTimeMillis() - 100));
        assertEquals("withdrawlFraudWarn", event.get("descr"));
    }

    private void sendWithdrawal(long acctNum, int amount)
    {
        Withdrawal event = new Withdrawal(acctNum, amount);
        epService.getEPRuntime().sendEvent(event);
    }

    private void sendFraudWarn(long acctNum, String text)
    {
        FraudWarning event = new FraudWarning(acctNum, text);
        epService.getEPRuntime().sendEvent(event);
    }
}
